README.md
# clojure-desafio-2

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-08-03
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-08-03
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-desafio-2/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-desafio-2/compare/0.1.0...0.1.1


test/clojure_desafio_2/core_test.clj
(ns clojure-desafio-2.core-test
  (:require [clojure.test :refer :all]
            [clojure-desafio-2.core :refer :all]))

test/clojure_desafio_2/logic_test.clj
(ns clojure-desafio-2.logic-test
  (:require [clojure.test :refer :all]
            [clojure-desafio-2.logic :refer :all]))

(use 'java-time)

(deftest adicionar-compra-test
  (testing "Testando o método de adicionar compra na lista de compras"

    (def lista-de-clientes [])
    (def lista-de-cartoes [])
    (def lista-de-compras [])

    (def lista-de-clientes (adicionar-cliente {:nome "Nyelson Barbosa", :cpf "22667862023", :email "nyelson.barbosa@nubank.com.br"} []))
    (def lista-de-cartoes (adicionar-cartao {:numero "5410133996442556", :cvv "984", :validade (local-date-time 2022 10 20), :limite 10000} []))

    (def cartao (get (seleciona-um-cartao "5410133996442556" lista-de-cartoes) 0))
    (def cartao (assoc-in cartao [:comprasRef] lista-de-compras))
    (def cliente (get (seleciona-um-cliente "22667862023" lista-de-clientes) 0))
    (def cliente (assoc-in cliente [:cartaoRef] cartao))

    (is (= 1
           (count (adicionar-compra {:data            (local-date-time 2021 10 20),
                                     :valor           100,
                                     :estabelecimento "Adidas",
                                     :categoria       "Vestuário"
                                     }
                                    lista-de-compras))
           ))

    (let [compra (peek (adicionar-compra {:data            (local-date-time 2021 10 20),
                                          :valor           1000,
                                          :estabelecimento "Adidas",
                                          :categoria       "Vestuário"
                                          }
                                         lista-de-compras))]
      (is (and (= 1000 (:valor compra)) (= "Adidas" (:estabelecimento compra)) (= "Vestuário" (:categoria compra)))))
    ))

(deftest listar-compras-por-cliente-test
  (testing "Testando método de listar compras por cliente"

    (def lista-de-clientes [])
    (def lista-de-cartoes [])
    (def lista-de-compras [])

    (def lista-de-clientes (adicionar-cliente {:nome "Nyelson Barbosa", :cpf "22667862023", :email "nyelson.barbosa@nubank.com.br"} []))
    (def lista-de-clientes (adicionar-cliente {:nome "Icaro Rios", :cpf "32257409000", :email "icaro.rios@nubank.com.br"} lista-de-clientes))

    ; Adicionando um cartão, sem associação a um cliente em específico e não validado - por enquanto
    (def lista-de-cartoes (adicionar-cartao {:numero "5410133996442556", :cvv "984", :validade (local-date-time 2022 10 20), :limite 10000} []))
    (def lista-de-cartoes (adicionar-cartao {:numero "5459268200059002", :cvv "737", :validade (local-date-time 2023 05 05), :limite 10000} lista-de-cartoes))

    ; Adicionando sete compras na lista de compras, para ser associada a algum cartão posteriormente, mas sem validações - por enquanto
    (def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 20), :valor 1000, :estabelecimento "Adidas", :categoria "Vestuário"} []))
    (def cartao (get (seleciona-um-cartao "5410133996442556" lista-de-cartoes) 0))
    (def cartao (assoc-in cartao [:comprasRef] lista-de-compras))
    (def cliente (get (seleciona-um-cliente "22667862023" lista-de-clientes) 0))
    (def cliente (assoc-in cliente [:cartaoRef] cartao))
    (def cartao2 (get (seleciona-um-cartao "5459268200059002" lista-de-cartoes) 0))
    (def cliente2 (get (seleciona-um-cliente "32257409000" lista-de-clientes) 0))
    (def cliente2 (assoc-in cliente2 [:cartaoRef] cartao2))

    (is (= [{:data (local-date-time 2021 10 20), :valor 1000, :estabelecimento "Adidas", :categoria "Vestuário"}]
           (lista-compras-por-cliente cliente)))

    (is (= 1
           (count (lista-compras-por-cliente cliente))))

    (is (nil? (lista-compras-por-cliente cliente2)))))

(deftest gastos-por-categoria-test
  (testing "Testando método de listar compras por cliente"

    (def lista-de-clientes [])
    (def lista-de-cartoes [])
    (def lista-de-compras [])

    (is (= (agrupa-categorias lista-de-compras) {}))

    (def lista-de-clientes (adicionar-cliente {:nome "Nyelson Barbosa", :cpf "22667862023", :email "nyelson.barbosa@nubank.com.br"} []))
    (def lista-de-clientes (adicionar-cliente {:nome "Icaro Rios", :cpf "32257409000", :email "icaro.rios@nubank.com.br"} lista-de-clientes))

    ; Adicionando um cartão, sem associação a um cliente em específico e não validado - por enquanto
    ; TO-DO: Colocar no formato de schema
    (def lista-de-cartoes (adicionar-cartao {:numero "5410133996442556", :cvv "984", :validade (local-date-time 2022 10 20), :limite 10000} []))
    (def lista-de-cartoes (adicionar-cartao {:numero "5459268200059002", :cvv "737", :validade (local-date-time 2023 05 05), :limite 10000} lista-de-cartoes))

    ; Adicionando sete compras na lista de compras, para ser associada a algum cartão posteriormente, mas sem validações - por enquanto
    ; TO-DO: Colocar no formato de schema
    (def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 20), :valor 1000, :estabelecimento "Adidas", :categoria "Vestuário"} []))
    (def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 21), :valor 250, :estabelecimento "Adidas", :categoria "Vestuário"} lista-de-compras))
    (def lista-de-compras (adicionar-compra {:data (local-date-time 2021 03 13), :valor 40, :estabelecimento "Burguer King", :categoria "Restaurante"} lista-de-compras))

    (def cartao (get (seleciona-um-cartao "5410133996442556" lista-de-cartoes) 0))
    (def cartao (assoc-in cartao [:comprasRef] lista-de-compras))
    (def cliente (get (seleciona-um-cliente "22667862023" lista-de-clientes) 0))
    (def cliente (assoc-in cliente [:cartaoRef] cartao))
    (def cartao2 (get (seleciona-um-cartao "5459268200059002" lista-de-cartoes) 0))
    (def cliente2 (get (seleciona-um-cliente "32257409000" lista-de-clientes) 0))
    (def cliente2 (assoc-in cliente2 [:cartaoRef] cartao2))

    (let [agrupamento {"Vestuário"   [{:data            (local-date-time 2021 10 20),
                                       :valor           1000,
                                       :estabelecimento "Adidas",
                                       :categoria       "Vestuário"}
                                      {:data            (local-date-time 2021 10 21),
                                       :valor           250,
                                       :estabelecimento "Adidas",
                                       :categoria       "Vestuário"}],
                       "Restaurante" [{:data            (local-date-time 2021 03 13),
                                       :valor           40,
                                       :estabelecimento "Burguer King",
                                       :categoria       "Restaurante"}]}]
      (is (= (agrupa-categorias lista-de-compras) agrupamento))
      )

    (def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 27), :valor 5, :estabelecimento "Padaria", :categoria "Alimentos" } lista-de-compras))

    (let [agrupamento {"Vestuário"   [{:data            (local-date-time 2021 10 20),
                                       :valor           1000,
                                       :estabelecimento "Adidas",
                                       :categoria       "Vestuário"}
                                      {:data            (local-date-time 2021 10 21),
                                       :valor           250,
                                       :estabelecimento "Adidas",
                                       :categoria       "Vestuário"}],
                       "Restaurante" [{:data            (local-date-time 2021 03 13),
                                       :valor           40,
                                       :estabelecimento "Burguer King",
                                       :categoria       "Restaurante"}]
                       "Alimentos" [{:data (local-date-time 2021 12 27),
                                     :valor 5,
                                     :estabelecimento "Padaria",
                                     :categoria "Alimentos"}]
                       }]
      (is (= (agrupa-categorias lista-de-compras) agrupamento))
      )

    ))

src/clojure_desafio_2/logic.clj
(ns clojure-desafio-2.logic
  (:require [clojure-desafio-2.credit-card :as w.db]))

(use 'java-time)

(defn agrupa-categorias
  "Agrupa e retorna uma lista de acordo com a categoria"
  [lista]
  (group-by :categoria lista))

(defn retorna-simbolo-por-categoria
  "Retorna o novo simbolo de acordo com a categoria passada e presente na lista agrupada"
  [lista categoria]
  (get lista categoria))

(defn somatorio-total
  "Retorna o somatório total dos valores passados pela lista no parâmetro"
  [lista]
  (->> lista (map :valor) (reduce +)))

(defn compras-por-estabelecimento-ou-valor
  [coluna filtro compra]
  (= filtro (coluna compra)))

(defn filtrando-compras
  [coluna filtro compras]
  (filter #(compras-por-estabelecimento-ou-valor coluna filtro %) compras))

(defn calcular-mes
  [inicial final compra]
  (println inicial final compra)
  (and (after? final (:data compra))
       (before? inicial (:data compra))))

(defn filtrando-compras-mes
  [inicial final compras]
  (filter #(calcular-mes inicial final %) compras))

(defn adicionar-compra
  "Adiciona uma compra na lista de compras"
  [compra lista]
  (conj lista compra))

(defn adicionar-cliente
  "Adiciona um cliente na lista de clientes"
  [cliente lista]
  (conj lista cliente))

(defn adicionar-cartao
  "Adiciona um cartão na lista de cartões"
  [cartao lista]
  (conj lista cartao))

(defn seleciona-um-cartao
  "Seleciona UM cartão no conjunto e variedades de cartões, pegando pelo número do cartão - que é único"
  [cartao lista]
  (filterv #(= cartao (:numero %)) lista))

(defn seleciona-um-cliente
  "Seleciona UM cliente no conjunto e variedades de clientes, pegando pelo número do cpf - que é único"
  [cliente lista]
  (filterv #(= cliente (:cpf %)) lista))

(defn lista-compras-por-cliente
  "Retorna lista de compras de acordo com o cliente passado por parâmetro"
  [cliente]
  (get (-> cliente :cartaoRef) :comprasRef))

; Definindo listas de clientes, cartões e de compras - antes da associação
; Adicionando dois clientes na lista de clientes, dados mockados e não validados - por enquanto
; TO-DO: Colocar no formato de schema


;(def lista-de-clientes (adicionar-cliente {:nome "Nyelson Barbosa", :cpf "22667862023", :email "nyelson.barbosa@nubank.com.br"} []))
;(def lista-de-clientes (adicionar-cliente {:nome "Icaro Rios",      :cpf "32257409000", :email "icaro.rios@nubank.com.br"     } lista-de-clientes))

; Adicionando um cartão, sem associação a um cliente em específico e não validado - por enquanto
; TO-DO: Colocar no formato de schema
;(def lista-de-cartoes (adicionar-cartao {:numero "5410133996442556", :cvv "984", :validade (local-date-time 2022 10 20), :limite 10000} []))
;(def lista-de-cartoes (adicionar-cartao {:numero "5459268200059002", :cvv "737", :validade (local-date-time 2023 05 05), :limite 10000} lista-de-cartoes))

; Adicionando sete compras na lista de compras, para ser associada a algum cartão posteriormente, mas sem validações - por enquanto
; TO-DO: Colocar no formato de schema
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 20), :valor 1000, :estabelecimento "Adidas",       :categoria "Vestuário"   } []))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 21), :valor 250,  :estabelecimento "Adidas",       :categoria "Vestuário"   } lista-de-compras))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 03 13), :valor 40,   :estabelecimento "Burguer King", :categoria "Restaurante" } lista-de-compras))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 05 29), :valor 300,  :estabelecimento "Nike",         :categoria "Vestuário"   } lista-de-compras))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 30), :valor 50,   :estabelecimento "Cinemark",     :categoria "Cinema"      } lista-de-compras))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 27), :valor 5,    :estabelecimento "Padaria",      :categoria "Alimentos"   } lista-de-compras))
;(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 25), :valor 70,   :estabelecimento "Boliche",      :categoria "Aleatório"   } lista-de-compras))
;(def cartao (get (seleciona-um-cartao "5410133996442556" lista-de-cartoes) 0))
;(def cartao (assoc-in cartao [:comprasRef] lista-de-compras))
;(def cliente (get (seleciona-um-cliente "22667862023" lista-de-clientes) 0))
;(def cliente (assoc-in cliente [:cartaoRef] cartao))
;(def cartao2 (get (seleciona-um-cartao "5459268200059002" lista-de-cartoes) 0))
;(def cliente2 (get (seleciona-um-cliente "32257409000" lista-de-clientes) 0))
;(def cliente2 (assoc-in cliente2 [:cartaoRef] cartao2))

src/clojure_desafio_2/prints.clj
(ns clojure-desafio-2.prints
  (:require [clojure-desafio-2.logic :refer :all]))

(use 'java-time)

(println "Início do projeto\n")

;(def lista-de-clientes [])
;(def lista-de-cartoes [])
;(def lista-de-compras [])

(def lista-de-clientes (adicionar-cliente {:nome "Nyelson Barbosa", :cpf "22667862023", :email "nyelson.barbosa@nubank.com.br"} []))
(def lista-de-clientes (adicionar-cliente {:nome "Icaro Rios",      :cpf "32257409000", :email "icaro.rios@nubank.com.br"     } lista-de-clientes))

;Adicionando um cartão, sem associação a um cliente em específico e não validado - por enquanto
;TO-DO: Colocar no formato de schema
(def lista-de-cartoes (adicionar-cartao {:numero "5410133996442556", :cvv "984", :validade (local-date-time 2022 10 20), :limite 10000} []))
(def lista-de-cartoes (adicionar-cartao {:numero "5459268200059002", :cvv "737", :validade (local-date-time 2023 05 05), :limite 10000} lista-de-cartoes))

;Adicionando sete compras na lista de compras, para ser associada a algum cartão posteriormente, mas sem validações - por enquanto
;TO-DO: Colocar no formato de schema
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 20), :valor 1000, :estabelecimento "Adidas",       :categoria "Vestuário"   } []))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 10 21), :valor 250,  :estabelecimento "Adidas",       :categoria "Vestuário"   } lista-de-compras))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 03 13), :valor 40,   :estabelecimento "Burguer King", :categoria "Restaurante" } lista-de-compras))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 05 29), :valor 300,  :estabelecimento "Nike",         :categoria "Vestuário"   } lista-de-compras))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 30), :valor 50,   :estabelecimento "Cinemark",     :categoria "Cinema"      } lista-de-compras))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 27), :valor 5,    :estabelecimento "Padaria",      :categoria "Alimentos"   } lista-de-compras))
(def lista-de-compras (adicionar-compra {:data (local-date-time 2021 12 25), :valor 70,   :estabelecimento "Boliche",      :categoria "Aleatório"   } lista-de-compras))
(def cartao (get (seleciona-um-cartao "5410133996442556" lista-de-cartoes) 0))
(def cartao (assoc-in cartao [:comprasRef] lista-de-compras))
(def cliente (get (seleciona-um-cliente "22667862023" lista-de-clientes) 0))
(def cliente (assoc-in cliente [:cartaoRef] cartao))
(def cartao2 (get (seleciona-um-cartao "5459268200059002" lista-de-cartoes) 0))
(def cliente2 (get (seleciona-um-cliente "32257409000" lista-de-clientes) 0))
(def cliente2 (assoc-in cliente2 [:cartaoRef] cartao2))

(println "Lista de clientes:" lista-de-clientes)
(println "Lista de cartões:" lista-de-cartoes)
(println "Lista de compras:" lista-de-compras)
(println "\n\n")

(println "Peguei o primeiro cartão:" (seleciona-um-cartao "5410133996442556" lista-de-cartoes))

; Foi utilizado o get cartão 0, para pegar o mapa referente ao cartão filtrado
(println "Associando as compras no primeiro cartão pego" (assoc-in cartao [:comprasRef] lista-de-compras))

(println "\n\nO primeiro cliente pesquisado foi:" (seleciona-um-cliente "22667862023" lista-de-clientes))

; Foi utilizada a mesma lógica do cartão para retornar o único cliente no vetor de cliente
(println "Associando o primeiro cartão ao primeiro cliente" (assoc-in cliente [:cartaoRef] cartao))

; Instanciando o segundo cartão e associando ao segundo cliente, sem adicionar nenhuma compra
(println "\nPeguei o segundo cartão:" (seleciona-um-cartao "5459268200059002" lista-de-cartoes))

(println "\n\nO segundo cliente pesquisado foi:" (seleciona-um-cliente "32257409000" lista-de-clientes))

(println "Associando o segundo cartão ao segundo cliente" (assoc-in cliente2 [:cartaoRef] cartao2))

(println "Listando compras do primeiro cliente:"  (lista-compras-por-cliente cliente))
(println "Listando compras do segundo cliente:"   (lista-compras-por-cliente cliente2))

(println "\nO total das compras referente a categoria vestuário é de:"  (somatorio-total (retorna-simbolo-por-categoria (agrupa-categorias lista-de-compras) "Vestuário")))
(println "O total das compras referente a categoria restaurante é de:"  (somatorio-total (retorna-simbolo-por-categoria (agrupa-categorias lista-de-compras) "Restaurante")))
(println "O total das compras é de:"                                    (somatorio-total lista-de-compras))

(println "\nRetornando o valor das comprar com o valor de 40:"  (->> lista-de-compras
                                                                     (filtrando-compras :valor 40)))

(println "Retornando o valor das comprar com o valor de 1000:"  (->> lista-de-compras
                                                                     (filtrando-compras :valor 1000)))

(println "Retornando o valor das comprar com o valor de 300:"   (->> lista-de-compras
                                                                     (filtrando-compras :valor 300)))

(println "\nRetornando o valor das compra na Adidas:"           (->> lista-de-compras
                                                                     (filtrando-compras :estabelecimento "Adidas")))

(println "Retornando o valor das compras no BK:"                (->> lista-de-compras
                                                                     (filtrando-compras :estabelecimento "Burguer King")))

(println "Retornando o valor das compra na Nike:"               (->> lista-de-compras
                                                                     (filtrando-compras :estabelecimento "Nike")))

(println "\nRetornando o valor do extrato no intervalo:"        (->> lista-de-compras
                                                                     (filtrando-compras-mes (local-date-time 2021 01 01 9 00) (local-date-time 2021 07 01 9 00)) somatorio-total))
